File: esw-dsl/src/main/scala/esw/dsl/script/LocationServiceDsl.scala
package esw.dsl.script

import csw.location.api.scaladsl.LocationService
import csw.location.models.ComponentType.Sequencer
import csw.location.models.{AkkaLocation, ComponentId}
import csw.location.models.Connection.AkkaConnection
import esw.dsl.Timeouts

import scala.async.Async.{async, await}
import scala.concurrent.{ExecutionContext, Future}

// fixme: implicit ec in script is strand ec, do we need that here?
//  can we use actor systems ec here and remove implicit ec from each api?
trait LocationServiceDsl {

  private[esw] val locationService: LocationService

  //todo: merge findSequencer and resolveSequencer
  def findSequencer(sequencerId: String, observingMode: String)(implicit ec: ExecutionContext): Future[AkkaLocation] =
    async {
      await(locationService.list)
      //fixme: sequencer has two registrations - http and akka, contains check will return any but we need akka here
        .find(location => location.connection.componentId.name.contains(s"$sequencerId@$observingMode"))
    }.collect {
      case Some(location: AkkaLocation) => location
      case Some(location) =>
        throw new RuntimeException(s"Sequencer is registered with wrong connection type: ${location.connection.connectionType}")
      case None => throw new IllegalArgumentException(s"Could not find any sequencer with name: $sequencerId@$observingMode")
    }

  def resolveSequencer(sequencerId: String, observingMode: String)(implicit ec: ExecutionContext): Future[AkkaLocation] =
    locationService
      .resolve(AkkaConnection(ComponentId(s"$sequencerId@$observingMode", Sequencer)), Timeouts.DefaultTimeout)
      .collect {
        case Some(location: AkkaLocation) => location
        case Some(location) =>
          throw new RuntimeException(s"Sequencer is registered with wrong connection type: ${location.connection.connectionType}")
        case None => throw new IllegalArgumentException(s"Could not find any sequencer with name: $sequencerId@$observingMode")
      }

}

File: script-dsl/src/main/kotlin/esw/ocs/dsl/highlevel/LocationServiceDsl.kt
package esw.ocs.dsl.highlevel

import csw.location.api.javadsl.ILocationService
import csw.location.models.AkkaLocation
import csw.location.models.Location
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.future.await

interface LocationServiceDsl : CoroutineScope {
    val locationService: ILocationService

    suspend fun resolveSequencer(sequencerId: String, observingMode: String): AkkaLocation {
        fun getSequencerLocation(locations: List<Location>): AkkaLocation {
            if (locations.isEmpty()) throw IllegalArgumentException("Could not find any sequencer with name: $sequencerId@$observingMode")
            val first = locations.component1()

            return when {
                first is AkkaLocation -> first
                locations.drop(1).isEmpty() -> throw RuntimeException("Sequencer is registered with wrong connection type: ${first.connection().connectionType()}")
                else -> getSequencerLocation(locations.drop(1))
            }
        }

        val sequencerLocations: List<Location> = locationService.list().await()
            .filter { it.connection().componentId().name().contains("$sequencerId@$observingMode") }

        return getSequencerLocation(sequencerLocations)
    }
}
